SQL: Disallow queries with inner LIMIT that cannot be executed in ES

Luegg opened the pull request:
Resolves https://github.com/elastic/elasticsearch/issues/75830

This PR only disallows subqueries with `LIMIT` if they cannot be translated into an semantically equivalent ES query. Thus, queries like `SELECT languages FROM (SELECT * FROM test_emp LIMIT 10)` that can be flattened correctly are still valid. Compared to disallowing all subqueries with `LIMIT`, this will only break queries that might return wrong results.

Additionally, this change also uncovered an issue with `TOP`. Because `TOP` is part of the `querySpecification` syntax, the `LogicalPlanBuilder` creates it as a child of `OrderBy` nodes. Thus, the query `SELECT TOP 10 * FROM test_emp ORDER BY salary` used to produce the logical plan
```
\ OrderBy
 \ Limit
  \ EsRelation[test_emp]
```
the semnatically correct plan would be 
```
\ Limit
 \ OrderBy
  \ EsRelation[test_emp]
```
This also had the effect that some queries with both `TOP` and `LIMIT` were not rejected by the parser.

Commit:
Disallow queries with inner limits that cannot be executed correctly

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

Commit:
Merge origin/master into fix/limitInInnerQueries

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java:
```diff
@@ -119,6 +124,21 @@ public LogicalPlan visitQueryNoWith(QueryNoWithContext ctx) {
+            if (unary.child() instanceof Limit) {
+                Limit limit = (Limit) unary.child();
+
+                return limit.replaceChildrenSameSize(Arrays.asList(
```
nit: `Collections#singletonList()` might work too.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/resources/select.sql-spec:
```diff
@@ -153,14 +153,10 @@ selectOrderByOrderByOrderByLimit
 SELECT * FROM (SELECT * FROM (SELECT * FROM test_emp ORDER BY emp_no DESC) ORDER BY emp_no ASC) ORDER BY emp_no DESC LIMIT 5;
 selectOrderByOrderByOrderByLimitLimit
 SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM test_emp ORDER BY emp_no DESC) ORDER BY emp_no ASC) ORDER BY emp_no DESC LIMIT 12) LIMIT 6;
-selectOrderByLimitSameOrderBy
-SELECT * FROM (SELECT * FROM (SELECT * FROM test_emp LIMIT 10) ORDER BY emp_no) ORDER BY emp_no LIMIT 5;
```
Wondering if this (type of) sequence of sub-selects couldn't be flattened safely. The outer ones don't change the ordering and out of multiple limitations, the smallest value could be determined (there's already code doing this comparison, IIRC?).

Luegg replied:
I think this particular query cannot be flattened as well because it also first limits and then sorts.

But I think you're right that there is a special case where the outer ORDER BY is redundant and could be eliminated as in `SELECT * FROM (SELECT * FROM test_emp ORDER BY emp_no LIMIT 10) ORDER BY emp_no`. This query should be equivalent to `SELECT * FROM (SELECT * FROM test_emp ORDER BY emp_no LIMIT 10)` which can safely be flattened.

I think the same principle applies for redundant filters as well: `SELECT * FROM (SELECT * FROM test_emp WHERE emp_no > 10 LIMIT 10) WHERE emp_no > 10`. And maybe also `GROUP BY`s.

I have the feeling this would best be implemented as an additional optimization though. A rule that eliminates redundant WHERE and ORDER BY.

Luegg replied:
I just noticed that in this particular case, the query could even be flattened because the order is not specified in the inner query. This would be a somewhat surprising behavior but probably aligned with SQL's "unordered bag of tuples" semantics.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/Verifier.java:
```diff
@@ -59,20 +63,19 @@
-                return;
-            }
-            if (p instanceof OrderExec) {
-                if (hasLimit.get() && orderBy.get() != null && ((OrderExec) p).order().equals(orderBy.get()) == false) {
```
Related to the removed query comment: can't we keep this "pass-through" check? I.e. chaining multiple orderings if they aren't combined with other clauses?

Commit:
Merge origin/master into fix/limitInInnerQueries

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/Verifier.java:
```diff
@@ -59,20 +63,19 @@
+        Set<LimitExec> limits = new TreeSet<>(Comparator.comparing(l -> l.source().text()));
+
+        plan.forEachDown(p -> {
+            if (p instanceof OrderExec || p instanceof FilterExec || p instanceof PivotExec || p instanceof AggregateExec) {
```
No need to do double iteration - the main forEach already traverses the tree so for each matching node doing another p.forEachDown is unnecessary.
Further more I don't think it's useful to the user to see all the instances where the same type of error occurs since a query with ORDER BY and a WHERE clause will report the error twice (once per entity).

I favor the initial approach since it's O(n) - traverse the tree once. (this could be optimized by ending the traversal on the first error however it's a rare enough case that we can ignore it).

Luegg replied:
👍 for the single iteration.

Also reporting only one error shouldn't make too much of a difference but I was more thinking about the cases where there is more than one offending `Limit` as in `SELECT * FROM (SELECT * FROM (SELECT * FROM test_emp LIMIT 10) LIMIT 20) WHERE languages > 5`. Since this seems to be a rather far-fetched case, the simpler one-error implementation should be fine.

costin reviewed (commented):
Left some comments on improving the complexity of the validation rule and the top parser.

Commit:
address review comments

Luegg commented:
@elasticmachine run elasticsearch-ci/bwc

Luegg commented:
@elasticmachine run elasticsearch-ci/bwc

Commit:
Merge origin/master into fix/limitInInnerQueries

Commit:
Merge origin/master into fix/limitInInnerQueries

Commit:
introduce PushDownAndCombineOrderBy optimization

Luegg commented:
@costin and @bpintea I've addressed your comments and had a look at what Bogdan mentioned about queries with redundant sort orders (`SELECT * FROM (SELECT * FROM test_emp ORDER BY emp_no LIMIT 10) ORDER BY emp_no`). In theory, we could allow such queries but I'm not sure it's worth it.

The last commit is an attempt to allow these queries and also to address https://github.com/elastic/elasticsearch/issues/76013 . Basically, it introduces a `PushDownAndCombineOrderBy` that folds multiple order bys if possible. The same logic could probably also be implemented in the `QueryFolder` but to me it looked more like an optimization.

Luegg commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -165,6 +167,9 @@ public LogicalPlan optimize(LogicalPlan verified) {
         );
 
+        Batch pushDownOrderBy = new Batch("Push Down Order By",
+            new PushDownAndCombineOrderBy());
```
Needs to run isolated from `PushDownAndCombineFilters` in order to avoid interfering push downs.

Luegg replied:
I think the optimizer would never converge to a stable plan in some cases as `OrderBy` and `Filter` would alternately be pushed down.

Luegg commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java:
```diff
@@ -701,6 +701,10 @@ protected PhysicalPlan rule(OrderExec plan) {
                 QueryContainer qContainer = exec.queryContainer();
 
+                if (qContainer.sort().isEmpty() == false) {
+                    throw new SqlIllegalArgumentException("QueryContainer already defines sort order");
```
I think we should be defensive here because the naive approach to just prepend additional sort orders leads to https://github.com/elastic/elasticsearch/issues/76013 .

Alternatively, `FoldOrderBy` (and `Verifier`) could take over the responsibility to merge `OrderBy`s from `PushDownAndCombineOrderBy`.

costin replied:
> Alternatively, FoldOrderBy (and Verifier) could take over the responsibility to merge OrderBys from PushDownAndCombineOrderBy.

I like this better.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1255,6 +1260,58 @@ private boolean foldable(Expression e) {
+            return plan;
+        }
+
+        private List<Order> combineOrders(List<Order> parentOrders, List<Order> childOrders) {
```
It's not clear to me how this method combines orders?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java:
```diff
@@ -102,10 +101,15 @@ public LogicalPlan visitQueryNoWith(QueryNoWithContext ctx) {
+            Source source = source(ctx.ORDER(), endContext);
+            List<Order> order = visitList(ctx.orderBy(), Order.class);
+
+            if (plan instanceof Limit) {
```
👍

costin reviewed (approved):
> The last commit is an attempt to allow these queries and also to address #76013 .

Please address that through a different PR (and +1 at enhancing FoldOrderBy instead of a separate rule) to keep things clean.
Otherwise LGTM.

Commit:
Revert "introduce PushDownAndCombineOrderBy optimization"

8d8d992c188948e040eb57e3fa743ee8ba804a58

Commit:
Merge origin/master into fix/limitInInnerQueries

Luegg commented:
I've reverted the last commit and will try to address https://github.com/elastic/elasticsearch/issues/76013 in a separate PR. This means that we will have a temporary regression as will no longer support queries with redundant sorts (e.g. `SELECT * FROM (SELECT * FROM test_emp ORDER BY emp_no LIMIT 10) ORDER BY emp_no`). Since it seems to be quite a special case, I would say that's justifiable.

Commit:
use replaceChild

elasticsearchmachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.14](https://github.com/elastic/elasticsearch/pull/76222)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/76222">](https://github.com/elastic/elasticsearch/pull/76222) |
| ✅ |  [7.x](https://github.com/elastic/elasticsearch/pull/76223)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/76223">](https://github.com/elastic/elasticsearch/pull/76223) |
